File: ios/MessagesExtension/Engine/Scoring.swift
import Foundation

// 1:1 port of src/scoring.ts. TTR point table, ticket connectivity, longest route.
struct FinalScore: Codable, Equatable {
    var routeScore: Int
    var ticketScore: Int
    var longestRoute: Int
    var longestBonus: Int
    var total: Int
}

enum Scoring {
    static let longestRouteBonus = 10
    private static let routeTable: [Int: Int] = [1: 1, 2: 2, 3: 4, 4: 7, 5: 10, 6: 15]

    static func routePoints(_ length: Int) -> Int { routeTable[length] ?? length }

    private static func playerRoutes(_ state: GameState, _ player: Int) -> [Route] {
        state.routes.filter { $0.claimedBy == player }
    }

    private static func adjacency(_ routes: [Route]) -> [Int: [(to: Int, route: Route)]] {
        var adj: [Int: [(to: Int, route: Route)]] = [:]
        for r in routes {
            adj[r.cityA, default: []].append((r.cityB, r))
            adj[r.cityB, default: []].append((r.cityA, r))
        }
        return adj
    }

    static func connected(_ routes: [Route], from: Int, to: Int) -> Bool {
        if from == to { return true }
        let adj = adjacency(routes)
        var seen: Set<Int> = [from]
        var stack = [from]
        while let city = stack.popLast() {
            for (next, _) in adj[city] ?? [] {
                if next == to { return true }
                if !seen.contains(next) { seen.insert(next); stack.append(next) }
            }
        }
        return false
    }

    static func ticketScore(_ state: GameState, _ player: Int) -> Int {
        let routes = playerRoutes(state, player)
        return state.players[player].tickets.reduce(0) { sum, t in
            sum + (connected(routes, from: t.cityA, to: t.cityB) ? t.points : -t.points)
        }
    }

    static func longestRoute(_ state: GameState, _ player: Int) -> Int {
        let adj = adjacency(playerRoutes(state, player))
        var best = 0
        var used: Set<Int> = []
        func dfs(_ city: Int, _ total: Int) {
            if total > best { best = total }
            for (to, route) in adj[city] ?? [] where !used.contains(route.id) {
                used.insert(route.id)
                dfs(to, total + route.length)
                used.remove(route.id)
            }
        }
        for city in adj.keys { dfs(city, 0) }
        return best
    }

    static func finalScores(_ state: GameState) -> [FinalScore] {
        let longest = state.players.indices.map { longestRoute(state, $0) }
        let maxLongest = longest.max() ?? 0
        let uniqueMax = longest.filter { $0 == maxLongest }.count == 1
        return state.players.indices.map { p in
            let routeScore = state.players[p].score
            let tScore = ticketScore(state, p)
            let bonus = (uniqueMax && longest[p] == maxLongest) ? longestRouteBonus : 0
            return FinalScore(routeScore: routeScore, ticketScore: tScore,
                              longestRoute: longest[p], longestBonus: bonus,
                              total: routeScore + tScore + bonus)
        }
    }

    static func finalWinner(_ state: GameState) -> Int? {
        let scores = finalScores(state)
        let maxTotal = scores.map(\.total).max() ?? 0
        let leaders = scores.indices.filter { scores[$0].total == maxTotal }
        return leaders.count == 1 ? leaders[0] : nil
    }
}

File: src/scoring.ts
import type { GameState, PlayerState, Route } from "./types.ts";

// Standard Ticket to Ride per-route point table (by length).
const ROUTE_POINTS: Record<number, number> = { 1: 1, 2: 2, 3: 4, 4: 7, 5: 10, 6: 15 };

export function routePoints(length: number): number {
  return ROUTE_POINTS[length] ?? length;
}

function playerRoutes(state: GameState, player: number): Route[] {
  return state.routes.filter((r) => r.claimedBy === player);
}

function adjacency(routes: Route[]): Map<number, { to: number; route: Route }[]> {
  const adj = new Map<number, { to: number; route: Route }[]>();
  const add = (from: number, to: number, route: Route) => {
    if (!adj.has(from)) adj.set(from, []);
    adj.get(from)!.push({ to, route });
  };
  for (const r of routes) {
    add(r.cityA, r.cityB, r);
    add(r.cityB, r.cityA, r);
  }
  return adj;
}

// Are the two cities connected through this player's claimed routes?
export function connected(routes: Route[], from: number, to: number): boolean {
  if (from === to) return true;
  const adj = adjacency(routes);
  const seen = new Set<number>([from]);
  const stack = [from];
  while (stack.length) {
    const city = stack.pop()!;
    for (const { to: next } of adj.get(city) ?? []) {
      if (next === to) return true;
      if (!seen.has(next)) { seen.add(next); stack.push(next); }
    }
  }
  return false;
}

// Tickets: + points if connected, - points if not (standard end scoring).
export function ticketScore(state: GameState, player: number): number {
  const routes = playerRoutes(state, player);
  return state.players[player].tickets.reduce(
    (sum, t) => sum + (connected(routes, t.cityA, t.cityB) ? t.points : -t.points),
    0,
  );
}

// Longest continuous path (trail: each segment used once), summed by length.
export function longestRoute(state: GameState, player: number): number {
  const routes = playerRoutes(state, player);
  const adj = adjacency(routes);
  let best = 0;
  const used = new Set<number>();

  const dfs = (city: number, total: number) => {
    if (total > best) best = total;
    for (const { to, route } of adj.get(city) ?? []) {
      if (used.has(route.id)) continue;
      used.add(route.id);
      dfs(to, total + route.length);
      used.delete(route.id);
    }
  };

  for (const city of adj.keys()) dfs(city, 0);
  return best;
}

export const LONGEST_ROUTE_BONUS = 10;

export interface FinalScore {
  routeScore: number;
  ticketScore: number;
  longestRoute: number;
  longestBonus: number;
  total: number;
}

export function finalScores(state: GameState): FinalScore[] {
  const longest = state.players.map((_, p) => longestRoute(state, p));
  const maxLongest = Math.max(...longest);

  return state.players.map((_, p) => {
    const routeScore = state.players[p].score;
    const tScore = ticketScore(state, p);
    // Every player tied for the longest path gets the bonus.
    const longestBonus = maxLongest > 0 && longest[p] === maxLongest ? LONGEST_ROUTE_BONUS : 0;
    return {
      routeScore,
      ticketScore: tScore,
      longestRoute: longest[p],
      longestBonus,
      total: routeScore + tScore + longestBonus,
    };
  });
}

// nil = tie, else winning player index.
export function finalWinner(state: GameState): number | null {
  const scores = finalScores(state);
  const max = Math.max(...scores.map((s) => s.total));
  const leaders = scores.flatMap((s, i) => (s.total === max ? [i] : []));
  return leaders.length === 1 ? leaders[0] : null;
}
